### PR TITLE
Update multi-task evaluation logic to avoid information leakage issue in lp and nfeat reconstruct task evaluation.

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
#789 

*Description of changes:*
Previously, in the eval() function of GSgnnMultiTaskLearningTrainer, both link prediction and node feature reconstruction tasks use the node embeddings computed with the entire graph. This will cause test edge leakage for link prediction tasks and target node node feature leakage for node feature reconstruction tasks. This PR fixes this issue.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/unit-tests/test_trainer.py [hunk 1]
```diff
@@ -417,7 +427,219 @@ def test_mtask_prepare_lp_mini_batch():
...
+        if self._test_results is not None:
+            compare_results(self._test_results, test_results)
+        return None, None
+
+    @property
+    def task_tracker(self):
+        return "dummy tracker"
+
+def test_mtask_eval():
+    tast_info_nc = TaskInfo(task_type=BUILTIN_TASK_NODE_CLASSIFICATION,
```
Comment (zhjwy9343) [inline tests/unit-tests/test_trainer.py:463]: tast -> task?

File: tests/unit-tests/test_trainer.py [hunk 2]
```diff
@@ -417,7 +427,219 @@ def test_mtask_prepare_lp_mini_batch():
...
+                            task_id='nc_task',
+                            task_config=None)
+    nc_dataloader = DummyGSgnnNodeDataLoader()
+    tast_info_nr = TaskInfo(task_type=BUILTIN_TASK_NODE_REGRESSION,
```
Comment (zhjwy9343) [inline tests/unit-tests/test_trainer.py:467]: ditto

File: tests/unit-tests/test_trainer.py [hunk 3]
```diff
@@ -417,7 +427,219 @@ def test_mtask_prepare_lp_mini_batch():
...
+                            task_id='nr_task',
+                            task_config=None)
+    nr_dataloader = DummyGSgnnNodeDataLoader()
+    tast_info_ec = TaskInfo(task_type=BUILTIN_TASK_EDGE_CLASSIFICATION,
```
Comment (zhjwy9343) [inline tests/unit-tests/test_trainer.py:471]: ditto

File: tests/unit-tests/test_trainer.py [hunk 4]
```diff
@@ -417,7 +427,219 @@ def test_mtask_prepare_lp_mini_batch():
...
+                            task_id='ec_task',
+                            task_config=None)
+    ec_dataloader = DummyGSgnnEdgeDataLoader()
+    tast_info_er = TaskInfo(task_type=BUILTIN_TASK_EDGE_REGRESSION,
```
Comment (zhjwy9343) [inline tests/unit-tests/test_trainer.py:475]: ditto

File: tests/unit-tests/test_trainer.py [hunk 5]
```diff
@@ -417,7 +427,219 @@ def test_mtask_prepare_lp_mini_batch():
...
+                            task_id='er_task',
+                            task_config=None)
+    er_dataloader = DummyGSgnnEdgeDataLoader()
+    task_config = GSConfig.__new__(GSConfig)
+    setattr(task_config, "train_mask", "train_mask")
+    tast_info_lp = TaskInfo(task_type=BUILTIN_TASK_LINK_PREDICTION,
```
Comment (zhjwy9343) [inline tests/unit-tests/test_trainer.py:481]: ditto

File: tests/unit-tests/test_trainer.py [hunk 6]
```diff
@@ -417,7 +427,219 @@ def test_mtask_prepare_lp_mini_batch():
...
+                            task_id='lp_task',
+                            task_config=task_config)
+
+    encoder_model = DummyGSgnnEncoderModel()
+    model = DummyGSgnnMTModel(encoder_model, decoders={tast_info_lp.task_id: "dummy"}, has_sparse=True)
```
Comment (zhjwy9343) [inline tests/unit-tests/test_trainer.py:486]: ditto

File: tests/unit-tests/test_trainer.py [hunk 7]
```diff
@@ -417,7 +427,219 @@ def test_mtask_prepare_lp_mini_batch():
...
+
+    mt_trainer = GSgnnMultiTaskLearningTrainer(model)
+    mt_trainer._device = 'cpu'
+
+    lp_dataloader = DummyGSgnnLinkPredictionDataLoader()
+    tast_info_nfr = TaskInfo(task_type=BUILTIN_TASK_RECONSTRUCT_NODE_FEAT,
```
Comment (zhjwy9343) [inline tests/unit-tests/test_trainer.py:492]: ditto

File: tests/unit-tests/test_trainer.py [hunk 8]
```diff
@@ -417,7 +427,219 @@ def test_mtask_prepare_lp_mini_batch():
...
+                            task_id='nfr_task',
+                            task_config=None)
+    nfr_dataloader = DummyGSgnnNodeDataLoader()
+    task_infos = [tast_info_nc, tast_info_nr, tast_info_ec,
```
Comment (zhjwy9343) [inline tests/unit-tests/test_trainer.py:496]: ditto

File: tests/unit-tests/test_trainer.py [hunk 9]
```diff
@@ -417,7 +427,219 @@ def test_mtask_prepare_lp_mini_batch():
...
+        val_loader = GSgnnMultiTaskDataLoader(None, task_infos, val_dataloaders)
+        test_loader = GSgnnMultiTaskDataLoader(None, task_infos, test_dataloaders)
+
+        target_res = {
+            "lp_task":lp_res,
+            "nfr_task":ntask_res
+        }
+        evaluator = MTaskCheckerEvaluator(target_res, target_res, 100)
+        mt_trainer.setup_evaluator(evaluator)
+        # test when val_loader is None
```
Comment (zhjwy9343) [inline tests/unit-tests/test_trainer.py:598]: May remove this incorrect comment line.

File: tests/unit-tests/test_trainer.py [hunk 10]
```diff
@@ -417,7 +427,219 @@ def test_mtask_prepare_lp_mini_batch():
...
+        target_res = {
+            "nc_task":ntask_res,
+            "nr_task":ntask_res,
+            "ec_task":etask_res,
+            "er_task":etask_res,
+            "nfr_task":ntask_res
+        }
+        evaluator = MTaskCheckerEvaluator(target_res, target_res, 200)
+        mt_trainer.setup_evaluator(evaluator)
+        # test when val_loader is None
```
Comment (zhjwy9343) [inline tests/unit-tests/test_trainer.py:617]: ditto

File: tests/unit-tests/test_trainer.py [hunk 11]
```diff
@@ -417,7 +427,219 @@ def test_mtask_prepare_lp_mini_batch():
...
+        target_res = {
+            "nc_task":ntask_res,
+            "nr_task":ntask_res,
+            "ec_task":etask_res,
+            "er_task":etask_res,
+            "lp_task":lp_res,
+        }
+        evaluator = MTaskCheckerEvaluator(target_res, target_res, 200)
+        mt_trainer.setup_evaluator(evaluator)
+        # test when val_loader is None
```
Comment (zhjwy9343) [inline tests/unit-tests/test_trainer.py:636]: ditto

File: python/graphstorm/run/gsgnn_mt/gsgnn_mt.py [hunk 12]
```diff
@@ -444,8 +444,12 @@ def main(config_args):
                 logging.warning("The training data do not have validation set.")
             if test_loader is None:
                 logging.warning("The training data do not have test set.")
-            task_evaluators[task.task_id] = \
-                create_evaluator(task)
+
+            if val_loader is None and test_loader is None:
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/run/gsgnn_mt/gsgnn_mt.py:448]: If no_validation is False, then we force users to give at least one loader for val or test for at least one task.

File: python/graphstorm/trainer/mt_trainer.py [hunk 13]
```diff
@@ -487,9 +489,9 @@ def eval(self, model, data, val_loader, test_loader, total_steps,
             The GNN model.
         data : GSgnnData
             The training dataset
-        val_loader: GSNodeDataLoader
+        mt_val_loader: GSNodeDataLoader
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/trainer/mt_trainer.py:492]: updated

File: python/graphstorm/trainer/mt_trainer.py [hunk 14]
```diff
@@ -506,42 +508,204 @@ def eval(self, model, data, val_loader, test_loader, total_steps,
...
+                """
+                if last_self_loop is False:
+                    model.gnn_encoder.skip_last_selfloop()
+                    new_embs = gen_embs()
+                    model.gnn_encoder.reset_last_selfloop()
+                    return new_embs
+                else:
+                    # if lask_self_loop is True
+                    # we can reuse the computed embs if any
+                    return embs if embs is not None else gen_embs()
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/trainer/mt_trainer.py:671]: Sending embs as an input arguement to gen_emb_for_nfeat_reconstruct is also confusing.
I added more comments to make the logic more clear.

File: tests/unit-tests/test_gnn.py [hunk 15]
```diff
@@ -2293,10 +2293,29 @@ def check_forward(mock_run_lp_mini_batch_predict,
...
-    check_forward()
+    gen_emb_for_nfeat_reconstruct(model, check_call_gen_embs)
+
+    call_self_loop = False
+    model = DummyGSgnnModel(encoder_model, has_sparse=False)
+    gen_emb_for_nfeat_reconstruct(model, check_call_gen_embs)
+
+    model = DummyGSgnnModel(None)
+    call_self_loop = True
+    def check_call_gen_embs(last_self_loop):
```
[the remark this answers is not in the document]
Comment (classicsong) [reply tests/unit-tests/test_gnn.py:2317]: Updated

**Review by zhjwy9343 — APPROVED** [review]
LGTM

**Pull request merged**